Create a blog on changes to 128-bit integers (#1281)
rust-lang/blog.rust-lang.org: posts/2024-00-00-i128-layout-update.md

---
layout: post
title: "Changes to `u128`/`i128` layout in 1.77 and 1.78"
author: Trevor Gross
team: Lang
---

> Rust has long had an inconsistency with C regarding the alignment of 128-bit integers.

Reviewer: ```suggestion
Rust has long had an inconsistency with C regarding the alignment of 128-bit integers on the x86-32 and x86-64 architectures.
```

This problem has recently been resolved, but the fix comes with some effects that are
worth being aware of.

As a user, you most likely do not need to worry about these changes unless you are:

1. Assuming the alignment of `i128`/`u128` rather than using `align_of`
1. Ignoring the `improper_ctypes*` lints and using these types in FFI

There are also no changes to architectures other than x86-32 and x86-64. If your
code makes heavy use of 128-bit integers, you may notice runtime performance increases
at a possible cost of additional memory use.

> This post is intended to clarify what changed, why it changed, and what to expect. If
> you are only looking for a compatibility matrix, jump to the
> [Compatibility](#compatibility) section.

Reviewer: ```suggestion
This post documents what the problem was, what changed to fix it, and what to expect. If
you're already familiar with the problem and you're only looking for a compatibility matrix, jump to the
[Compatibility](#compatibility) section.
```

# Background

Data types have two intrinsic values that relate to how they can be arranged in memory;
size and alignment. A type's size is the amount of space it takes up in memory, and its
alignment specifies which addresses it is allowed to be placed at.

The size of simple types like primitives is usually unambiguous, being the exact size of
the data they represent with no padding (unused space). For example, an `i64` always has
a size of 64 bits or 8 bytes.

> Alignment, however, can seem less consistent. An 8-byte integer _could_ reasonably be

Reviewer: ```suggestion
Alignment, however, can seem less consistent. An 8-byte integer _could_ be
```

Offset of b (u16) in Foo: 2
Alignment of Foo: 2
Offset of b (u64) in Bar: 8
Alignment of Bar: 8
```

We see that within a struct, a type will always be placed such that its offset is a
multiple of its alignment.

These numbers are not arbitrary; the application binary interface (ABI) says what they
should be. In the x86-64 [psABI] (processor-specific ABI) for System V (Unix & Linux),
_Figure 3.1: Scalar Types_ tells us exactly how primitives should be represented:

| C type           | Rust equivalent | `sizeof` | Alignment (bytes) |
| ---------------- | --------------- | -------- | ----------------- |
| `char`           | `i8`            | 1        | 1                 |
| `unsigned char`  | `u8`            | 1        | 1                 |
| `short`          | `i16`           | 2        | 2                 |
| `unsigned short` | `u16`           | 2        | 2                 |
| `long`           | `i64`           | 8        | 8                 |
| `unsigned long`  | `u64`           | 8        | 8                 |

The ABI only specifies C types, but Rust follows the same definitions both for
compatibility and for the performance benefits.

# The Incorrect Alignment Problem

> It is easy to imagine that if two implementations disagree on the alignment of a data
> type, they would not be able to reliably share data containing that type. Well...

Reviewer: ```suggestion
If two implementations disagree on the alignment of a data
type, they cannot reliably share data containing that type. Rust had inconsistent alignment for 128-bit types:
```

```rust=
println!("alignment of i128: {}", align_of::<i128>());
```

```text=
// rustc 1.76.0
alignment of i128: 8
```

```c=
printf("alignment of __int128: %zu\n", _Alignof(__int128));
```

```text=
// gcc 13.2
alignment of __int128: 16

// clang 17.0.1
alignment of __int128: 16
```

> Looks like Rust disagrees![^align-godbolt] Looking back at the [psABI], we can see that
> Rust indeed is in the wrong here:

Reviewer: ```suggestion
Looking back at the [psABI], we can see that Rust has the wrong alignment here:
```

| C type              | Rust equivalent | `sizeof` | Alignment (bytes) |
| ------------------- | --------------- | -------- | ----------------- |
| `__int128`          | `i128`          | 16       | 16                |
| `unsigned __int128` | `u128`          | 16       | 16                |

It turns out this isn't because of something that Rust is actively doing incorrectly:
layout of primitives comes from the LLVM codegen backend used by both Rust and Clang,
among other languages, and it has the alignment for `i128` hardcoded to 8 bytes.

Clang does not have this issue only because of a workaround, where the alignment is
manually set to 16 bytes before handing the type to LLVM. This fixes the layout issue
but has been the source of some other minor problems.[^f128-segfault][^va-segfault]
Rust does no such manual adjustement, hence the issue reported at
<https://github.com/rust-lang/rust/issues/54341>.

# The Calling Convention Problem

> It happens that there an additional problem: LLVM does not always do the correct thing

Reviewer: ```suggestion
There's an additional problem: LLVM does not always do the correct thing
```

| Rust ≥ 1.77 with LLVM \< 18        | GCC (any version)   | Storage compatible, has calling bug |
| Rust ≥ 1.77 with LLVM \< 18        | Clang (any version) | Storage compatible, has calling bug |
| Rust \< 1.77[^l]                   | GCC (any version)   | Incompatible                        |
| Rust \< 1.77[^l]                   | Clang (any version) | Incompatible                        |
| GCC (any version)                  | Clang ≥ 18          | Fully compatible                    |
| GCC (any version)                  | Clang \< 18         | Storage compatible with calling bug |

[^l]: Rust < 1.77 with LLVM 18 will have some degree of compatibility, this is just
      an uncommon combination.

# Effects & Future Steps

As mentioned in the introduction, most users will see no effects of this change
unless you are already doing something questionable with these types.

Starting with Rust 1.77, it will be reasonably safe to start experimenting with
128-bit integers in FFI, with some more certainty coming with the LLVM update
in 1.78. There is [ongoing discussion] about lifting the lint in an upcoming

> version, but it remains to be seen when that will actually happen.

Reviewer: ```suggestion
version, but we want to be cautious and avoid introducing silent breakage
for users whose Rust compiler may be built with an older LLVM.
```

[relevance to Rust was discovered]: https://github.com/rust-lang/rust/issues/54341#issuecomment-1064729606
[initial performance run]: https://github.com/rust-lang/rust/pull/116672/#issuecomment-1858600381
[known issue in llvm]: https://github.com/llvm/llvm-project/issues/41784
[psabi]: https://www.uclibc.org/docs/psABI-x86_64.pdf
[ongoing discussion]: https://github.com/rust-lang/lang-team/issues/255

> [^align-godbolt]: https://godbolt.org/z/h94Ge1vMW

Reviewer: ```suggestion
[align-godbolt]: https://godbolt.org/z/h94Ge1vMW
```
